File: c_leetcode/maxProduct.c
#include "stdio.h"
#include <assert.h>

int maxProduct(int* nums, int numsSize)
{
    int i = 0, j = numsSize - 1;
    int ans = 0;

    while (i != j) {
        int tmp = (nums[i] - 1) * (nums[j] - 1);
        if (ans < tmp) {
            ans = tmp;
        }
        if (nums[i] > nums[j]) {
            j--;
        } else {
            i++;
        }
    }
    return ans;
}

int main()
{
    int nums1[4] = { 3, 4, 5, 2 };
    assert(maxProduct(nums1, 4) == 12);

    int nums2[4] = { 1, 5, 4, 5 };
    assert(maxProduct(nums2, 4) == 16);

    int nums3[2] = { 3, 7 };
    assert(maxProduct(nums3, 2) == 12);

    return 0;
}

File: rust_leetcode/src/bin/max_product.rs
struct Solution;
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0, nums.len() - 1);
        let mut ans = 0;

        while left != right {
            let tmp = (nums[left] - 1) * (nums[right] - 1);
            if tmp > ans {
                ans = tmp;
            }
            if nums[left] > nums[right] {
                right -= 1;
            } else {
                left += 1;
            }
        }
        ans
    }
}

fn main() {
    println!("{}", Solution::max_product(vec![3, 4, 5, 2]));
}

#[cfg(test)]
mod tests {
    use crate::Solution;
    #[test]
    fn test() {
        assert_eq!(Solution::max_product(vec![3, 4, 5, 2]), 12);
        assert_eq!(Solution::max_product(vec![1, 5, 4, 5]), 16);
        assert_eq!(Solution::max_product(vec![3, 7]), 12);
    }
}
